File: src/nautilus_ctp_adapter/runtime/market.py
from __future__ import annotations

from dataclasses import dataclass, field

from .models import CtpRuntimeCommand, CtpRuntimeCommandKind


def _subscription_key(venue_symbol: str | None, exchange_id: str | None) -> str | None:
    if not venue_symbol:
        return None
    return venue_symbol if not exchange_id else f"{exchange_id}:{venue_symbol}"


@dataclass(slots=True)
class CtpMarketRuntime:
    _subscriptions: set[str] = field(default_factory=set)

    def on_command(self, command: CtpRuntimeCommand) -> None:
        key = _subscription_key(command.venue_symbol, command.exchange_id)
        if key is None:
            return

        if command.kind is CtpRuntimeCommandKind.SUBSCRIBE_MARKET_DATA:
            self._subscriptions.add(key)
        elif command.kind is CtpRuntimeCommandKind.UNSUBSCRIBE_MARKET_DATA:
            self._subscriptions.discard(key)

    @property
    def subscription_count(self) -> int:
        return len(self._subscriptions)

    def is_subscribed(self, venue_symbol: str, exchange_id: str | None = None) -> bool:
        key = _subscription_key(venue_symbol, exchange_id)
        return False if key is None else key in self._subscriptions

    def snapshot(self) -> tuple[str, ...]:
        return tuple(sorted(self._subscriptions))

File: rust/ctp_runtime_core/src/market.rs
use std::collections::BTreeSet;

use crate::commands::{CtpCommand, CtpCommandKind};

fn subscription_key(venue_symbol: &Option<String>, exchange_id: &Option<String>) -> Option<String> {
    match (venue_symbol.as_deref(), exchange_id.as_deref()) {
        (Some(symbol), Some(exchange)) => Some(format!("{exchange}:{symbol}")),
        (Some(symbol), None) => Some(symbol.to_owned()),
        _ => None,
    }
}

#[derive(Debug, Clone, Default)]
pub struct CtpMarketRuntime {
    subscriptions: BTreeSet<String>,
}

impl CtpMarketRuntime {
    pub fn on_command(&mut self, command: &CtpCommand) {
        let Some(key) = subscription_key(&command.venue_symbol, &command.exchange_id) else {
            return;
        };

        match command.kind {
            CtpCommandKind::SubscribeMarketData => {
                self.subscriptions.insert(key);
            }
            CtpCommandKind::UnsubscribeMarketData => {
                self.subscriptions.remove(&key);
            }
            _ => {}
        }
    }

    pub fn subscription_count(&self) -> usize {
        self.subscriptions.len()
    }

    pub fn is_subscribed(&self, venue_symbol: &str, exchange_id: Option<&str>) -> bool {
        let key = match exchange_id {
            Some(exchange) => format!("{exchange}:{venue_symbol}"),
            None => venue_symbol.to_owned(),
        };
        self.subscriptions.contains(&key)
    }
}
